parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,633 @@
+
+			// initialization logic: we know that this runs exactly once every block,
+			// which means we can put the initialization logic here to remove the
+			// sequencing problem.
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				if vfp.relay_chain_height >= apply_block {
+					PendingValidationFunction::kill();
+					Self::put_parachain_code(&validation_function);
+					Self::deposit_event(Event::ValidationFunctionApplied(vfp.relay_chain_height));
+				}
+			}
+
+			storage::unhashed::put(VALIDATION_FUNCTION_PARAMS, &vfp);
+			DidUpdateVFPs::put(true);
+			<T::OnValidationFunctionParams as OnValidationFunctionParams>::on_validation_function_params(vfp);
+		}
+
+		fn on_finalize() {
+			assert!(DidUpdateVFPs::take(), "VFPs must be updated once per block");
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get validation function parameters.
+	fn validation_function_params() -> ValidationFunctionParams {
+		// this storage value is set by cumulus during block validation,
+		// and also by the inherent from this module.
+		storage::unhashed::get(VALIDATION_FUNCTION_PARAMS)
+			.expect("validation function params must be set")
+	}
+
+	/// Get validation function parameters.
+	///
+	/// This will return `None` if this module's inherent has not yet run.
+	/// If it returns `Some(_)`, the validation function params are current for this block.
+	pub fn get_validation_function_params() -> Option<ValidationFunctionParams> {
+		if DidUpdateVFPs::try_get().unwrap_or(false) {
+			Some(Self::validation_function_params())
+		} else {
+			None
+		}
+	}
+
+	/// Put a new validation function into a particular location where polkadot
+	/// monitors for updates. Calling this function notifies polkadot that a new
+	/// upgrade has been scheduled.
+	fn notify_polkadot_of_pending_upgrade(code: &[u8]) {
+		storage::unhashed::put_raw(NEW_VALIDATION_CODE, code);
+	}
+
+	/// Put a new validation function into a particular location where this
+	/// parachain will execute it on subsequent blocks.
+	fn put_parachain_code(code: &[u8]) {
+		storage::unhashed::put_raw(well_known_keys::CODE, code);
+	}
+
+	/// `true` when a code upgrade is currently legal
+	pub fn can_set_code() -> bool {
+		Self::validation_function_params()
+			.code_upgrade_allowed
+			.is_some()
+	}
+
+	/// The maximum code size permitted, in bytes.
+	pub fn max_code_size() -> u32 {
+		Self::validation_function_params().max_code_size
+	}
+}
+
+impl<T: Trait> ProvideInherent for Module<T> {
+	type Call = Call<T>;
+	type Error = sp_inherents::MakeFatalError<()>;
+	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+		// If the inherent is not present, this returns None early. This in turn will
+		// cause the on_finalize assertion to fail.
+		let vfp: ValidationFunctionParams = data.get_data(&INHERENT_IDENTIFIER).ok().flatten()?;
```

Comment:
I will change that, but I'd like to talk with you at some point about what the expected semantics are in `create_inherent`, then, because they clearly aren't what I imagined.